py/picca/io.py

```diff
@@ -1375,116 +1112,111 @@ def read_from_minisv_desi(in_dir,
-                    hdul['{}_WAVELENGTH'.format(spectrograph)].read())
-                spec['FL'] = hdul['{}_FLUX'.format(spectrograph)].read()
-                spec['IV'] = (
-                    hdul['{}_IVAR'.format(spectrograph)].read() *
-                    (hdul['{}_MASK'.format(spectrograph)].read()==0))
+                spec['log_lambda'] = np.log10(hdul[f'{color}_WAVELENGTH'].read())
+                spec['FL'] = hdul[f'{color}_FLUX'].read()
+                spec['IV'] = (hdul[f'{color}_IVAR'].read() *
+                             (hdul[f'{color}_MASK'].read()==0))
                 w = np.isnan(spec['FL']) | np.isnan(spec['IV'])
                 for key in ['FL','IV']:
                     spec[key][w] = 0.
-                spec['RESO'] = hdul['{}_RESOLUTION'.format(
-                    spectrograph)].read()
-                spec_data[spectrograph]=spec
+                spec['RESO'] = hdul[f'{color}_RESOLUTION'].read()
+                spec_data[color] = spec
             except OSError:
-                userprint(f"error when reading {spectrograph}-band data")
+                userprint(f"ERROR: when reading {color}-band data")
 
         hdul.close()
         plate_spec = int(f"{tile_spec}{petal_spec}")
 
-        select=(plate==plate_spec)&(night==night_spec)
-        print('\nThis is tile {}, petal {}, night {}'.format(tile_spec,petal_spec,night_spec))
-        thingid_qsos = thingid[select]
-        plate_qsos = plate[select]
-        night_qsos = night[select]
-        fiberid_qsos = fiberid[select]
-        for t, p, m, f in zip(thingid_qsos,plate_qsos,night_qsos,fiberid_qsos):
-            w_t = in_thingids == t
-            if w_t.sum() == 0:
-                print(f"\nError reading thingid {t}\n")
-                print(f"spectra : {spec}\n")
-                print(f"plate_spec : {plate_spec}\n")
-                continue
+        select = ( (catalog['TILEID'] == tile_spec) &
+                   (catalog['PETAL_LOC'] == petal_spec) & 
+                   (catalog['NIGHT']  == night_spec) )
+        userprint(f'This is tile {tile_spec}, petal {petal_spec}, night {night_spec}')
+
+        #-- Loop over quasars in catalog inside this tile-petal
+        for entry in catalog[select]:
 
-            forest=None
+            #-- Find which row in tile contains this quasar
+            w_t = np.where(targetid_spec == entry['TARGETID'])[0][0]
+
+            #-- Loop over three spectrograph arms and coadd fluxes
+            forest = None
             for spec in spec_data.values():
-                ivar = spec['IV'][w_t]
-                flux = (ivar*spec['FL'][w_t]).sum(axis=0)
-                ivar = ivar.sum(axis=0)
-                w = ivar > 0.
-                flux[w] /= ivar[w]
+                ivar = spec['IV'][w_t]*1
+                flux = spec['FL'][w_t]*1
 
                 if pk1d is not None:
-                    reso_sum = spec['RESO'][w_t].sum(axis=0)
+                    reso_sum = spec['RESO'][w_t]*1
```

Comment:
why the "*1"? I don't think this is the same operation.

Reply:
I want to make sure we are doing copies of the data. 
Also I didn't understand the operations with ivar (some weighted average).

Reply:
Ok, I think the sum was used here to do an average over all selected spectra (in case one has the same targetid in the sample mulitple times), that converts an [nspec, ndiag, npix] array to an [ndiag, npix] array so cannot be removed here. Also the *1 for a copy is not really good style ;-) why not use .copy() instead

All the ivar operations are to do inverse variance weighted coadding between different spectra with same THINGID/TARGETID within the same coadd-file. The actual coadd routine, currently only coadds the different arms I think.

All of those routines probaby need a bit of an update to be more clear...

Reply:
In this part of the code, there cannot be two spectra for one single object/targetid. I think the sum was to convert an 2D array into a 1D array. 
I use np.where instead of a mask array so this sum() is not needed. I tested and it works.

Reply:
Ok, I only had this ever do the 2d to 1d conversion (it definitely does so with current data). Personally I like the mask array syntax more than np.where, but that's probably a question of taste.
If I'm not misreading the code having two spectra of the same target on the same tile in the same night would lead to coadding them in the old version, but would just pick the first of them in the new. Given that the outputs we're using are already coadded this should never happen. Might still be worth putting in a check and issuing a warning/raising an exception if there's ever more than one index in w_t.

I'm still against the "*1" as copy() is clearer to the reader (but I do not think the copying is even necessary in this case), i.e. a future maintainer might be tempted to remove a "*1" assuming it does nothing.

Reply:
I've added the warning and added .copy() for each array.

Reply:
This is addressed with commit d464ea8